### ISSUE TITLE
Miscompile with -fglobal-isel, -ftrivial-auto-var-init=pattern, and thinlto

### ISSUE THREAD

**Issue by nico** [llvm/llvm-project#59376]
Repro:

1. Download and gunzip attachment at https://bugs.chromium.org/p/chromium/issues/detail?id=1383873#c53
2.
```shell
bin/clang++ -fno-delete-null-pointer-checks -fno-ident -fno-strict-aliasing -fstack-protector -fcolor-diagnostics -fmerge-all-constants -ffp-contract=off -flto=thin -fsplit-lto-unit -fwhole-program-vtables -fcomplete-member-pointers -arch arm64 -ffile-compilation-dir=. -no-canonical-prefixes -ftrivial-auto-var-init=pattern -fno-omit-frame-pointer -mmacos-version-min=10.13 -O2 -std=c++20 -fno-exceptions -fno-rtti repro.ii -shared -o librepro.dylib -fuse-ld=lld -Wl,-undefined,dynamic_lookup -w -isysroot $(xcrun -show-sdk-path) -fno-global-isel
```
3.
```shell
bin/clang++ -fno-delete-null-pointer-checks -fno-ident -fno-strict-aliasing -fstack-protector -fcolor-diagnostics -fmerge-all-constants -ffp-contract=off -flto=thin -fsplit-lto-unit -fwhole-program-vtables -fcomplete-member-pointers -arch arm64 -ffile-compilation-dir=. -no-canonical-prefixes -ftrivial-auto-var-init=pattern -fno-omit-frame-pointer -mmacos-version-min=10.13 -O2 -std=c++20 -fno-exceptions -fno-rtti repro.ii -shared -o librepro.dylib -fuse-ld=lld -Wl,-undefined,dynamic_lookup -w -isysroot $(xcrun -show-sdk-path) -fglobal-isel
```
4. Compare disassembly for function `__ZN5blink22NGFragmentItemsBuilder17ConvertToPhysicalERKNS_12PhysicalSizeE`

(The only difference between commands 2 and 3 is `-fno-global-isel` vs `-fglobal-isel`.)

The good dylib has this sequence:

```console
% otool -tV librepro.good.dylib | rg -A 233 __ZN5blink22NGFragmentItemsBuilder17ConvertToPhysicalERKNS_12PhysicalSizeE: | rg -A 3 csinc
0000000000002f80	csinc	w8, w8, wzr, ne
0000000000002f84	strh	w8, [sp, #0x38]
0000000000002f88	str	wzr, [sp, #0x40]
0000000000002f8c	ldr	x0, [x24]
```

The bad dylib has this sequence:

```console
% otool -tV librepro.bad.dylib | rg -A 233 __ZN5blink22NGFragmentItemsBuilder17ConvertToPhysicalERKNS_12PhysicalSizeE: | rg -A 3 csinc
0000000000003110	csinc	w8, w8, wzr, ne
0000000000003114	strh	w8, [sp, #0x30]
0000000000003118	mov	w9, #-0x55555556
000000000000311c	str	x9, [sp, #0x30]
```

In the bad dylib, we store something to `sp+0x30`, and then clobber it with `0xaaaaaaaa` (the `-ftrivial-auto-var-init=pattern` pattern) immediately.

This is reduced from https://crbug.com/1383873

**Comment by nico** [comment]
@aemerson

**Comment by aemerson** [comment]
Seems to be a bug in the store-merging pass (LoadStoreOpt.cpp).

**Comment by aemerson** [comment]
I wasn't able to see the exact same code as you were seeing after disassembling that dylib, but after `-emit-llvm` and running it through llc, I was able to spot a bug which looks to be the same as the one you discussed in the chromium bug report. A test case to demonstrate the issue

```
; Here store of 5 and 9 can be merged, others have aliasing barriers.
define i32 @test_alias_3xs16(ptr %ptr, ptr %ptr2, ptr %ptr3, ptr noalias %safe_ptr) {
  %safeld = load i32, ptr %safe_ptr
  %addr2 = getelementptr i32, ptr %ptr, i64 1
  store i32 5, ptr %addr2
  store i16 0, ptr %addr2 ; aliases directly with store above.
  %addr3 = getelementptr i32, ptr %ptr, i64 2
  store i32 9, ptr %addr3
  %addr4 = getelementptr i32, ptr %ptr, i64 3
  store i32 14, ptr %addr4
  ret i32 %safeld
}
```

There's a very specific set of circumstances that triggers this (the presence of an unrelated non-aliasing operation, in this case the `%safeld` load, is needed. The current llc will generate:
```
_test_alias_3xs16:                      ; @test_alias_3xs16
	.cfi_startproc
; %bb.0:
	mov	x9, #5
	mov	x8, x0
	ldr	w0, [x3]
	movk	x9, #9, lsl #32
	mov	w10, #14
	strh	wzr, [x8, #4]
	stur	x9, [x8, #4]
	str	w10, [x8, #12]
	ret
```
...which merges the stores of 5 and 9, critically merging across the aliasing 16 bit store of 0. My local build with the fix generates:
```
_test_alias_3xs16:                      ; @test_alias_3xs16
	.cfi_startproc
	mov	x10, #9
	mov	x8, x0
	ldr	w0, [x3]
	mov	w9, #5
	movk	x10, #14, lsl #32
	str	w9, [x8, #4]
	strh	wzr, [x8, #4]
	str	x10, [x8, #8]
	ret
```
as expected.

**Comment by aemerson** [comment]
Fixed in 66d64aac36a6

Thanks very much @nico for the report. I appreciate it wasn't easy to debug and get a repro.
